jmslib/src/main/java/com/redhat/mqe/lib/SenderClient.java

```diff
@@ -103,7 +103,23 @@ public void startClient() {
                 }
 
                 // Send messages
-                msgProducer.send(message);
+                try {
+                    msgProducer.send(message);
+                } catch (Exception e) {
+                    if (e.getMessage().endsWith("released by receiver")) {
```

Comment:
@fgiorgetti is there any docs for this "release" scenario?

Reply:
There are some upstream references, like:

> If the connection to the broker fails, Dispatch Router automatically attempts to reestablish the connection and reroute message deliveries to any available alternate destinations. However, some deliveries could be returned to the sender with a RELEASED or MODIFIED disposition. Therefore, you should ensure that your clients can handle these deliveries appropriately (generally by resending them).

Source: http://qpid.apache.org/releases/qpid-dispatch-1.9.0/user-guide/index.html

Reply:
Is this the only way to detect a RELEASED disposition in qpid-jms? Do you really have to match on the text of the exception? If yes, then I suppose we can merge this.

Reply:
The exception was a JMSException and I was also in doubt if matching the text was a good idea. I will do some local testing and see if there is a better way to match other than going through the text. I will update you guys once I have news.

Reply:
@jdanekrh @michalxo I have also found that the string from "getCause()" contains the original exception class, so we can also match it with:
`e.getCause().toString().contains("org.apache.qpid.jms.provider.exceptions.ProviderDeliveryReleasedException")`

Reply:
Or we can do `instanceof` check against `e.getCause()`, I think. Is the cause an instance of `ProviderDeliveryReleasedException`?

Reply:
I prefer instanceof Exception. That's the cleanest. Message might change. Exception as well though :D

Reply:
Ok, I haven't done that because I wanted to avoid adding extra dependency to jmslib's pom.xml.
So I added it and then we can check instanceof properly, see:

```
diff --git a/jmslib/pom.xml b/jmslib/pom.xml
index 7142f69..ca841ce 100644
--- a/jmslib/pom.xml
+++ b/jmslib/pom.xml
@@ -30,6 +30,10 @@
         <relativePath>../parent/pom.xml</relativePath>
     </parent>
 
+    <properties>
+        <qpid.jms.client.version>0.48.0</qpid.jms.client.version>
+    </properties>
+
     <artifactId>jmslib</artifactId>
 
     <dependencies>
@@ -50,6 +54,12 @@
             <version>1.2.2-SNAPSHOT</version>
         </dependency>
 
+        <dependency>
+            <groupId>org.apache.qpid</groupId>
+            <artifactId>qpid-jms-client</artifactId>
+            <version>${qpid.jms.client.version}</version>
+        </dependency>
+
         <!--scope provided means that it is a compileOnly dependency-->
         <dependency>
             <groupId>org.apache.geronimo.specs</groupId>
```

Reply:
Right, I forgot we then end up with that dependency. Yeah, we don't want the dependency. We do not want to have qpid-jms in e.g. the core client. So, what are the options?

* check exception message string, we've almost done it that way in another case, https://github.com/rh-messaging/cli-java/pull/77/files
* use reflection (get the exception class name as a string and compare on that string)
* do do this [wait for it]

Reply:
3 is hardcore, 2 is rather nice and not too crazy, I think, 1 is what we started with.

Reply:
@fgiorgetti so please do either reflection or go back to the string comparison, either is ok with me. Again, I forgot that we cannot do instanceof, because the send code is shared for all jms libs.

Reply:
@jdanekrh alright, I will see if the reflection will end up looking fine (without adding too much complexity) and then I will update the PR (removing the new dep).

Reply:
The following code snippet worked just fine, so I am going with it:
```
                    switch (e.getCause().getClass().getName()) {
                        case "org.apache.qpid.jms.provider.exceptions.ProviderDeliveryReleasedException":
                            String onRelease = senderOptions.getOption(ClientOptions.ON_RELEASE).getValue();
                            LOG.trace(String.format("Message released [action: %s]", onRelease));
                            switch (onRelease) {
                                case "fail":
                                    throw e;
                                case "retry":
                                    continue;
                            }
                        default:
                            // Preserve original behavior
                            throw e;
                    }
```